consensus/src/dag/types.rs

```diff
@@ -0,0 +1,282 @@
+        Self { node, certificate }
+    }
+}
+
+impl Deref for CertifiedNode {
+    type Target = Node;
+
+    fn deref(&self) -> &Self::Target {
+        &self.node
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct NodeDigestSignature {
+    epoch: u64,
+    digest: HashValue,
+    signature: bls12381::Signature,
+}
+
+impl DAGMessage for Node {
+    fn epoch(&self) -> u64 {
+        self.metadata.epoch
+    }
+}
+
+impl DAGMessage for NodeDigestSignature {
+    fn epoch(&self) -> u64 {
+        self.epoch
+    }
+}
+
+impl DAGMessage for NodeCertificate {
+    fn epoch(&self) -> u64 {
+        self.epoch
+    }
+}
+
+pub struct SignatureBuilder {
+    digest: HashValue,
+    partial_signatures: PartialSignatures,
+    epoch_state: Arc<EpochState>,
+}
+
+impl SignatureBuilder {
+    pub fn new(digest: HashValue, epoch_state: Arc<EpochState>) -> Self {
+        Self {
+            digest,
+            partial_signatures: PartialSignatures::empty(),
+            epoch_state,
+        }
+    }
+}
+
+impl BroadcastStatus for SignatureBuilder {
+    type Ack = NodeDigestSignature;
+    type Aggregated = NodeCertificate;
+    type Message = Node;
+
+    fn add(&mut self, peer: Author, ack: Self::Ack) -> anyhow::Result<Option<Self::Aggregated>> {
+        ensure!(self.digest == ack.digest, "Digest mismatch");
+        self.partial_signatures.add_signature(peer, ack.signature);
+        Ok(self
+            .epoch_state
+            .verifier
+            .check_voting_power(self.partial_signatures.signatures().keys())
+            .ok()
+            .map(|_| {
+                let aggregated_signature = self
+                    .epoch_state
+                    .verifier
+                    .aggregate_signatures(&self.partial_signatures)
+                    .expect("Signature aggregation should succeed");
+                NodeCertificate {
+                    epoch: self.epoch_state.epoch,
+                    digest: self.digest,
+                    signatures: aggregated_signature,
+                }
+            }))
+    }
+}
+
+pub struct CertificateAckState {
+    num_validators: usize,
+    received: HashSet<Author>,
+}
+
+impl CertificateAckState {
+    pub fn new(num_validators: usize) -> Self {
+        Self {
+            num_validators,
+            received: HashSet::new(),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct CertifiedAck {
+    epoch: u64,
+}
+
+impl DAGMessage for CertifiedAck {
+    fn epoch(&self) -> u64 {
+        self.epoch
+    }
+}
+
+impl BroadcastStatus for CertificateAckState {
+    type Ack = CertifiedAck;
+    type Aggregated = ();
+    type Message = NodeCertificate;
+
+    fn add(&mut self, peer: Author, _ack: Self::Ack) -> anyhow::Result<Option<Self::Aggregated>> {
+        self.received.insert(peer);
+        if self.received.len() == self.num_validators {
+            Ok(Some(()))
+        } else {
+            Ok(None)
+        }
+    }
+}
```

Comment:
nit: move this impl BroadcastStatus above to consolidate same type implementations?